core/services/blockheaderfeeder/block_header_feeder.go

```diff
@@ -0,0 +1,255 @@
+}
+
+// NewBlockHeaderFeeder creates a new BlockHeaderFeeder instance.
+func NewBlockHeaderFeeder(
+	logger logger.Logger,
+	coordinator blockhashstore.Coordinator,
+	bhs blockhashstore.BHS,
+	batchBHS BatchBHS,
+	blockHeaderProvider BlockHeaderProvider,
+	waitBlocks int,
+	lookbackBlocks int,
+	latestBlock func(ctx context.Context) (uint64, error),
+	gethks keystore.Eth,
+	getBlockhashesBatchSize uint16,
+	storeBlockhashesBatchSize uint16,
+	fromAddresses []ethkey.EIP55Address,
+	chainID *big.Int,
+) *BlockHeaderFeeder {
+	return &BlockHeaderFeeder{
+		lggr:                      logger,
+		coordinator:               coordinator,
+		bhs:                       bhs,
+		batchBHS:                  batchBHS,
+		waitBlocks:                waitBlocks,
+		lookbackBlocks:            lookbackBlocks,
+		latestBlock:               latestBlock,
+		stored:                    make(map[uint64]struct{}),
+		lastRunBlock:              0,
+		getBlockhashesBatchSize:   getBlockhashesBatchSize,
+		storeBlockhashesBatchSize: storeBlockhashesBatchSize,
+		blockHeaderProvider:       blockHeaderProvider,
+		gethks:                    gethks,
+		fromAddresses:             fromAddresses,
+		chainID:                   chainID,
+	}
+}
+
+// BlockHeaderFeeder checks recent VRF coordinator events and stores any blockhashes for blocks within
+// waitBlocks and lookbackBlocks that have unfulfilled requests.
+type BlockHeaderFeeder struct {
+	lggr                      logger.Logger
+	coordinator               blockhashstore.Coordinator
+	bhs                       blockhashstore.BHS
+	batchBHS                  BatchBHS
+	waitBlocks                int
+	lookbackBlocks            int
+	latestBlock               func(ctx context.Context) (uint64, error)
+	stored                    map[uint64]struct{}
+	blockHeaderProvider       BlockHeaderProvider
+	lastRunBlock              uint64
+	getBlockhashesBatchSize   uint16
+	storeBlockhashesBatchSize uint16
+	gethks                    keystore.Eth
+	fromAddresses             []ethkey.EIP55Address
+	chainID                   *big.Int
+}
+
+// Run the feeder.
+func (f *BlockHeaderFeeder) Run(ctx context.Context) error {
+	latestBlockNumber, err := f.latestBlock(ctx)
+	if err != nil {
+		f.lggr.Errorw("Failed to fetch current block number", "error", err)
+		return errors.Wrap(err, "fetching block number")
+	}
+
+	fromBlock, toBlock := blockhashstore.GetSearchWindow(int(latestBlockNumber), f.waitBlocks, f.lookbackBlocks)
+	if toBlock == 0 {
+		// Nothing to process, no blocks are in range.
+		return nil
+	}
+
+	lggr := f.lggr.With("latestBlock", latestBlockNumber, "fromBlock", fromBlock, "toBlock", toBlock)
+	lggr.Debug("searching for unfulfilled blocks")
+
+	blockToRequests, err := blockhashstore.GetUnfulfilledBlocksAndRequests(ctx, lggr, f.coordinator, fromBlock, toBlock)
+	if err != nil {
+		return err
+	}
+
+	minBlockNumber := f.findLowestBlockNumberWithoutBlockhash(ctx, lggr, blockToRequests)
```

Comment:
It seems like there isn't any difference going for the oldest missed request vs the newest missed request, except that the oldest ensures we don't lose any by falling beyond the lookback window. Just thinking out loud.